src/contracts/protocol/libraries/logic/LiquidationLogic.sol

```diff
@@ -318,127 +514,152 @@ library LiquidationLogic {
     DataTypes.ReserveCache memory debtReserveCache,
-    DataTypes.ExecuteLiquidationCallParams memory params,
-    uint256 healthFactor
-  ) internal view returns (uint256, uint256) {
-    uint256 userVariableDebt = IERC20(debtReserveCache.variableDebtTokenAddress).balanceOf(
-      params.user
-    );
+    DataTypes.ReserveData storage debtReserve,
+    DataTypes.UserConfigurationMap storage userConfig,
+    address user,
+    address debtAsset,
+    uint256 userReserveDebt,
+    uint256 actualDebtToLiquidate,
+    bool hasNoCollateralLeft
+  ) internal {
+    // Prior v3.1, there were cases where, after liquidation, the `isBorrowing` flag was left on
+    // even after the user debt was fully repaid, so to avoid this function reverting in the `_burnScaled`
+    // (see ScaledBalanceTokenBase contract), we check for any debt remaining.
+    if (userReserveDebt != 0) {
+      debtReserveCache.nextScaledVariableDebt = IVariableDebtToken(
+        debtReserveCache.variableDebtTokenAddress
+      ).burn(
+          user,
+          hasNoCollateralLeft ? userReserveDebt : actualDebtToLiquidate,
+          debtReserveCache.nextVariableBorrowIndex
+        );
+    }
 
-    uint256 closeFactor = healthFactor > CLOSE_FACTOR_HF_THRESHOLD
-      ? DEFAULT_LIQUIDATION_CLOSE_FACTOR
-      : MAX_LIQUIDATION_CLOSE_FACTOR;
+    uint256 outstandingDebt = userReserveDebt - actualDebtToLiquidate;
+    if (hasNoCollateralLeft && outstandingDebt != 0) {
+      /**
+       * Special handling of GHO. Implicitly assuming that virtual acc !active == GHO, which is true.
+       * Scenario 1: The amount of GHO debt being liquidated is greater or equal to the GHO accrued interest.
+       *             In this case, the outer handleRepayment will clear the storage and all additional operations can be skipped.
+       * Scenario 2: The amount of debt being liquidated is lower than the GHO accrued interest.
+       *             In this case handleRepayment will be called with the difference required to clear the storage.
+       *             If we assume a liquidation of n debt, and m accrued interest, the difference is k = m-n.
+       *             Therefore we call handleRepayment(k).
+       *             Additionally, as the dao (GHO issuer) accepts the loss on interest on the bad debt,
+       *             we need to discount k from the deficit (via reducing outstandingDebt).
+       * Note: If a non GHO asset is liquidated and GHO bad debt is created in the process, Scenario 2 applies with n = 0.
+       */
+      if (!debtReserveCache.reserveConfiguration.getIsVirtualAccActive()) {
+        uint256 accruedInterest = IGhoVariableDebtToken(debtReserveCache.variableDebtTokenAddress)
+          .getBalanceFromInterest(user);
+        // handleRepayment() will first discount the protocol fee from an internal `accumulatedDebtInterest` variable
+        // and then burn the excess GHO
+        if (accruedInterest != 0 && accruedInterest > actualDebtToLiquidate) {
```

Comment:
maybe I m a bit blind, but `handleRepayment` won't be executed in scenario 2 because of this conditional, no?